ui/components/app/snaps/snap-permissions-list/snap-permissions-list.js

```diff
@@ -1,52 +1,128 @@
-import React from 'react';
+import React, { useState, useEffect, useMemo } from 'react';
 import PropTypes from 'prop-types';
 import { useSelector } from 'react-redux';
-import { getWeightedPermissions } from '../../../../helpers/utils/permission';
 import { useI18nContext } from '../../../../hooks/useI18nContext';
-import { Box } from '../../../component-library';
+import { Box, ButtonLink } from '../../../component-library';
 import {
   getMultipleTargetsSubjectMetadata,
   getSnapsMetadata,
 } from '../../../../selectors';
+import {
+  Display,
+  FlexDirection,
+  JustifyContent,
+} from '../../../../helpers/constants/design-system';
+import SnapPermissionAdapter from '../snap-permission-adapter';
+import { PermissionWeightThreshold } from '../../../../../shared/constants/permissions';
 import { getSnapName } from '../../../../helpers/utils/util';
-import SnapPermissionCell from '../snap-permission-cell';
+import { getWeightedPermissions } from '../../../../helpers/utils/permission';
 
 export default function SnapPermissionsList({
   snapId,
   snapName,
   permissions,
   connections,
   showOptions,
+  turnOffAbstraction,
+  showAllPermissions,
 }) {
   const t = useI18nContext();
-  const snapsMetadata = useSelector(getSnapsMetadata);
-  const permissionsToShow = {
-    ...permissions,
-    connection_permission: connections ?? {},
-  };
+
+  const combinedPermissions = useMemo(() => {
+    return { ...permissions, connection_permission: connections ?? {} };
+  }, [permissions, connections]);
+
   const targetSubjectsMetadata = useSelector((state) =>
     getMultipleTargetsSubjectMetadata(state, connections),
   );
 
-  return (
-    <Box className="snap-permissions-list">
-      {getWeightedPermissions({
+  const snapsMetadata = useSelector(getSnapsMetadata);
+
+  const [showAll, setShowAll] = useState(false);
+
+  const [permissionsToDisplay, setPermissionsToDisplay] = useState([]);
+
+  const [weightedPermissions, setWeightedPermissions] = useState({});
+
+  useEffect(() => {
+    let finalPermissions = weightedPermissions;
+    if (Object.keys(finalPermissions).length === 0) {
+      finalPermissions = getWeightedPermissions({
         t,
-        permissions: permissionsToShow,
+        permissions: combinedPermissions,
         subjectName: snapName,
         getSubjectName: getSnapName(snapsMetadata),
-      }).map((permission, index) => (
-        <SnapPermissionCell
+      });
+      setWeightedPermissions(finalPermissions);
+    }
+
+    if (turnOffAbstraction && !showAll) {
+      setShowAll(true);
+    }
+
+    if (Object.keys(finalPermissions).length <= 3 && !showAll) {
```

Comment:
The logic for when to start abstracting permissions away should be documented with comments.

Potentially the `3` here should be moved to a constant as well and explained.

Reply:
Added constant where other related stuff is defined. Explained with the comment above it. Also added some comments. Let me know if you would like something more, I'll add.